src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -125,12 +125,19 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
 	private void populateCacheTask(MessageCacheItem cacheItem) {
 		Callable<MessageCacheItem> callable = () -> {
     		try {
-    			// Pass cacheItem to getMessages so that:
-				// 1. A previously stored etag, if any, can be used for the next HTTP request.
-				// 2. CacheItem properties such as etag, timestamp and maxAgeMillis can be refreshed 
-				// 	 with new properties from the next HTTP response.
-				fetchMessages(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
-    			return cacheItem;
+    			/* Pass cacheItem to fetchMessages so that:
+					1. A previously stored etag, if any, can be used for the next HTTP request.
+					2. CacheItem properties such as etag, timestamp and maxAgeMillis can be refreshed
+						with new values from the next HTTP response.
+				*/
+				String cacheItemLocale = cacheItem.getLocale();
+    			if (cacheItemLocale.equals(this.dto.getLocale())) {
+					fetchMessages(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+				} else {  //If cacheItem's locale is not the requested locale, it means it is for a fallback locale. Hence, refresh the appropriate fallback locale's cache
+					MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), cacheItemLocale, dto.getProductID(), dto.getVersion());
+					new ComponentService(fallbackLocaleDTO).fetchMessages(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
```

Comment:
For some reasons, the requested locale failed last time. We still need to try the requested locale first, then fallback to default locale if failure.

Reply:
I will update line 103 with this: `if (cacheItem.isExpired() || !cacheItem.getLocale().equals(this.dto.getLocale()))`

Reply:
I'm afraid it can't. If one locale gets the default locale's messages, it will always get the default locale's messages. It's wrong.

Reply:
Right! My mistake. See the changes please

Reply:
I mean that the requested locale fails not because locale isn't supported but other reasons such as network error. So next time to update cache, still need to get the requested locale first.

Reply:
Yes, that is what is happening as I explained to Huihui. Let's connect over Slack

Reply:
I have also updated the code to make it easier to understand, but same logic.